jax/experimental/pjit.py

```diff
@@ -850,18 +845,16 @@ def keep_where(l, should_keep):
       *(pv.get_known() for pv in in_pvals if pv.is_known()),
       **known_params)
   if num_residuals:
-    known_out_vals, residual_vals = split_list(all_known_outs, [-num_residuals])
+    known_out_vals, residual_vals = \
+        split_list(all_known_outs, [len(all_known_outs) - num_residuals])
   else:
     known_out_vals, residual_vals = all_known_outs, ()
-  known_tracers_out = [trace.new_const(known_out) for known_out in known_out_vals]
   residual_tracers = [trace.new_instantiated_const(residual) for residual in residual_vals]
 
-  # Prepare the unknown jaxpr
-  # TODO(apaszke): map_jaxpr will break caching!
-  unknown_jaxpr = raw_unknown_jaxpr.map_jaxpr(lambda jaxpr: pe._drop_vars(
-      jaxpr,
-      drop_ins=known_ins + (False,) * num_residuals,
-      drop_outs=known_outs))
+  # The convention of partial_eval_jaxpr_nounits is to place residual binders
+  # at the front of the jaxpr produced, so we move them here.
```

Comment:
Maybe a tiny bit more here, e.g.:
```python
# The convention of partial_eval_jaxpr_nounits is to place residual binders
# at the front of the jaxpr produced, so we move them to the back since later
# we'll pass the residual tracers into the primitive after the unknown tracers.
```

Reply:
SG!